### ISSUE TITLE
Authentication problems

### ISSUE THREAD

**Issue by nclemeur** [DavidFW1960/Aussie-Broadband-Usage-Meter#1]
I am experiencing problems to authenticate. I think it can come from 2 sources:

o the encryption/decryption mechanism does work properly for complicated passwords
o (quite likely) the parameters send in the request are not URL encoded

**Comment by DavidFW1960** [comment]
I have a pretty complicated password and have not seen any issue. When I changed my password at Aussie it didn't like that much though..... Ha.

The password is stored in an encrypted text file at C:\Users\user-name-here\AppData\Roaming\Rainmeter-ABB
Maybe try closing rainmeter and deleting the files in there and restarting it.

My password is upper, lower, numbers and a special character and is 16 digits long. I have had issues with characters in other programs that don't like 'reserved' characters and need them to be escaped (like < or >). You also might have put in the wrong password when you initially entered it? If you want to email me your password I can try changing my password at Aussie to that and testing it there. david@wilfam.net

**Comment by nclemeur** [comment]
I am pretty sure my password was correct. I tried twice by deleting the files containing the password. I also tried to return the password as is from the Decrypt function, but it still did not work. So I just used a browser and recorded the value that was sent for my password (in the network log) and the password is definitely URL encoded. So I just copied the Url encoded version and hard coded it in the usage script and it worked.

To give you an idea, my password is along the line of:

RM]%9^t('+>K3!dC^zW(:

So in summary, I am pretty sure it is broken. The encrypt/decrypt function might work correctly, but the url encoding is missing (unless you are doing it the encrypt function, and then the encrypt/decrypt mechanism is broken).

**Comment by DavidFW1960** [comment]
OK I changed my password to your 'idea' above, deleted my config files and refreshed the skin. Entered the new username and password and it is working as normal. Can't find an issue.

EDIT: Scratch that. It doesn't work. Not sure what I can do about that.

**Comment by DavidFW1960** [comment]
Hmm..... I might post a job on upwork to see if someone can debug the encrypt/de-encrypt function but of course I'm not selling this usage meter. I'll see if I can get it fixed economically.

**Comment by nclemeur** [comment]
I manage to get it to work by simply adding a `URLEncode ` function. Seems that the decrypt/encrypt function work.

Then you need to use that function right after you have decrypted the password.

https://www.experts-exchange.com/questions/26732683/UrlEncode-Using-Wscript.html

```
Function URLEncode(StringToEncode)
  Dim TempAns, CurChr, iChar
  CurChr = 1
  Do Until CurChr - 1 = Len(StringToEncode)
    iChar = Asc(Mid(StringToEncode, CurChr, 1))
    If (iChar > 47 And iChar < 58)  Or (iChar > 64 And iChar < 91) Or (iChar > 96 And iChar < 123) Then
      TempAns = TempAns & Mid(StringToEncode, CurChr, 1)
    ElseIf iChar = 32 Then
      TempAns = TempAns & "%" & Hex(32)      
    Else
      TempAns = TempAns & "%" & Right("00" & Hex(Asc(Mid(StringToEncode, CurChr, 1))), 2)
    End If
    CurChr = CurChr + 1
  Loop
  URLEncode = TempAns
End Function
```

**Comment by DavidFW1960** [comment]
So - if I open up the ABB-usage.vbs and insert that function at line 222 is that all I need to do? Do I need to edit the values being parsed in that function at all? Do I need to make any other changes to the script so that function is called? Are you able to email me the whole vbs file you are using? (that is working)

I did actually post a job for this on upwork but if you've got a working solution I'll update the script for the benefit of everyone.

**Comment by nclemeur** [comment]
Just send you a pull request

**Comment by DavidFW1960** [comment]
ok if all went well I just merged it. I'll download it and see how we go here and then I'll update the rmskin

**Comment by DavidFW1960** [comment]
It doesn't like chr on line 217 Invalid Procedure Call or Argument

**Comment by nclemeur** [comment]
Sorry I did not do it properly or the merge did not work... But the lines
106 and 107 should actually be after reading the password (ie before line 115
[Set wxml = CreateObject(..])

Cheers

**Comment by nclemeur** [comment]
Just created another pull request that should fix it

**Comment by DavidFW1960** [comment]
ok! no error now. Just trying to change my aussie password to test that...... Thanks. Fingers crossed.
